kedro/io/core.py

```diff
@@ -71,10 +70,17 @@ class DatasetAlreadyExistsError(DatasetError):
     pass
 
 
-class DataSetAlreadyExistsError(metaclass=DeprecatedClassMeta):
-    # pylint: disable=missing-class-docstring, too-few-public-methods
-
-    _DeprecatedClassMeta__alias = DatasetAlreadyExistsError
+def __getattr__(name):
+    if name in _DEPRECATED_ERROR_CLASSES:
+        alias = name.replace("DataSet", "Dataset")
+        warnings.warn(
+            f"{name} has been renamed to {alias}, and the alias will be removed in "
+            f"Kedro 0.19.0",
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        return globals()[alias]
```

Comment:
What about having a mapping between old names new objects so that we don't have to use the `globals()` mechanism? Something like

```python
_DEPRECATED_ERROR_CLASSES = {
    "DataSetError": DatasetError,
    "DataSetNotFoundError": DatasetNotFoundError,
    "DataSetAlreadyExistsError": DatasetAlreadyExistsError,
}
```

the `if name in _DEPRECATED_ERROR_CLASSES` would still work, and then we could do `return _DEPRECATED_ERROR_CLASSES[name]`.

Reply:
The code with `globals()` pretty much comes from the PEP that proposed adding module `__getattr__`: https://peps.python.org/pep-0562/

But I don't know why it shouldn't work to use a dictionary here, so I can try if `globals()` is bad to use. My understanding was that this may be a good use of `globals()`.

I can also see how other libraries using `__getattr__` for depreciation do it, potentially.

Reply:
`globals()` removed in https://github.com/kedro-org/kedro/pull/2673/commits/9b13345b52b9c23baea74b7f24aa96493ebfdd57 (following your suggestion).

About the `exec()`, I'm not sure what to do/haven't looked into it more. Open to suggestions; also OK to ignore it for now since it's only in test code that can be removed post-0.19.0-release.

Reply:
Actually, I will try overwriting `__dir__` and see if it happens to help.